tasmota/tasmota_xdrv_driver/xdrv_65_tuyamcubr.ino

```diff
@@ -0,0 +1,916 @@
+		return (TUYAMCUBR_P_START);
+
+	case TUYAMCUBR_P_SKIP:
+		if (++p->p_off >= p->p_len)
+			return (TUYAMCUBR_P_SKIP_CKSUM);
+		return (nstate);
+	case TUYAMCUBR_P_SKIP_CKSUM:
+		return (TUYAMCUBR_P_START);
+	}
+
+	p->p_sum += byte;
+
+	return (nstate);
+}
+
+static uint8_t
+tuyamcubr_write(struct tuyamcubr_softc *sc, const void *data, size_t len)
+{
+	TasmotaSerial *serial = sc->sc_serial;
+	const uint8_t *bytes = (const uint8_t *)data;
+	uint8_t cksum = 0;
+	size_t i;
+
+	for (i = 0; i < len; i++) {
+		uint8_t b = bytes[i];
+		serial->write(b);
+		cksum += b;
+	}
+
+	return (cksum);
+}
+
+static void
+tuyamcubr_send(struct tuyamcubr_softc *sc, uint8_t command,
+    const void *data, size_t len)
+{
+	TasmotaSerial *serial = sc->sc_serial;
+	struct tuyamcubr_header h = {
+		.header = { TUYAMCUBR_H_ONE, TUYAMCUBR_H_TWO },
+		.version = 0x00,
+		.command = command,
+		.datalen = htons(len),
+	};
+	uint8_t cksum = 0;
+
+	if (len) {
+		AddLog(LOG_LEVEL_DEBUG,
+		    TUYAMCUBR_FMT("send version 0x%02x command 0x%02x: %*_H"),
+		    h.version, h.command, len, data);
+	} else {
+		AddLog(LOG_LEVEL_DEBUG,
+		    TUYAMCUBR_FMT("send version 0x%02x command 0x%02x"),
+		    h.version, h.command);
+	}
+
+	cksum += tuyamcubr_write(sc, &h, sizeof(h));
+	if (len > 0)
+		cksum += tuyamcubr_write(sc, data, len);
+	cksum = tuyamcubr_cksum_fini(cksum);
+	serial->write(cksum);
+	serial->flush();
+}
+
+/* if we have polymorphic funcions then we may as well (ab)use them */
+static void
+tuyamcubr_send(struct tuyamcubr_softc *sc, uint8_t command)
+{
+	tuyamcubr_send(sc, command, NULL, 0);
+}
+
+static void
+tuyamcubr_heartbeat(struct tuyamcubr_softc *sc, unsigned int deadline)
+{
+	sc->sc_deadline += deadline;
+	tuyamcubr_send(sc, TUYAMCUBR_CMD_HEARTBEAT);
+}
+
+static struct tuyamcubr_dp *
+tuyamcubr_find_dp(struct tuyamcubr_softc *sc, uint32_t index, uint8_t type)
+{
+	struct tuyamcubr_dp *dp;
+
+	if (index > 0xff)
+		return (NULL);
+
+	STAILQ_FOREACH(dp, &sc->sc_dps, dp_entry) {
+		if (dp->dp_id == index &&
+		    dp->dp_type == type)
+			return (dp);
+	}
+
+	return (NULL);
+}
+
+static void
+tuyamcubr_cmnd_data(struct tuyamcubr_softc *sc, uint8_t type)
+{
+	const struct tuyamcubr_data_type *dt = &tuyamcubr_data_types[type];
+	struct {
+		struct tuyamcubr_data_header h;
+		uint8_t value[4]; /* only up to 4 bytes */
+	} data;
+	size_t len = sizeof(data.h) + dt->t_len;
+	struct tuyamcubr_dp *dp;
+
+	dp = tuyamcubr_find_dp(sc, XdrvMailbox.index, type);
+	if (dp == NULL) {
+		ResponseCmndChar_P(PSTR("Unknown DpId"));
+		return;
+	}
+
+	if (XdrvMailbox.data_len == 0) {
+		ResponseCmndNumber(dp->dp_value);
+		return;
+	}
+
+	if (XdrvMailbox.payload < 0x00 || XdrvMailbox.payload > dt->t_max) {
+		ResponseCmndChar_P(PSTR("Invalid"));
+		return;
+	}
+
+	dp->dp_value = XdrvMailbox.payload;
+
+	data.h.dpid = dp->dp_id;
+	data.h.type = dp->dp_type;
+	data.h.len = htons(dt->t_len);
+	dt->t_wr(data.value, dp->dp_value);
+
+	tuyamcubr_send(sc, TUYA_CMD_SET_DP, &data, len);
+
```

Comment:
Pls change TUYA_CMD_SET_DP to TUYAMCUBR_CMD_SET_DP

Reply:
done